File: Data Structures & Algorithms/implement-prefix-tree/submission-1.java
class PrefixTree {
    private PrefixTree[] children;
    private boolean isEnd;

    public PrefixTree() {
        children = new PrefixTree[26];
        isEnd = false;
    }

    public void insert(String word) {
        PrefixTree node = this;
        for (char c : word.toCharArray()) {
            int i = c - 'a';
            if (node.children[i] == null) {
                node.children[i] = new PrefixTree();
            }
            node = node.children[i];
        }
        node.isEnd = true;
    }

    public boolean search(String word) {
        PrefixTree node = find(word);
        return node != null && node.isEnd;
    }

    public boolean startsWith(String prefix) {
        return find(prefix) != null;
    }

    private PrefixTree find(String word) {
        PrefixTree node = this;
        for (char c : word.toCharArray()) {
            int i = c - 'a';
            if (node.children[i] == null) {
                return null;
            }
            node = node.children[i];
        }
        return node;
    }
}

File: Data Structures & Algorithms/design-word-search-data-structure/submission-1.ts
class TrieNode {
    children: Map<string, TrieNode> = new Map();
    isEnd: boolean = false;
}

class WordDictionary {
    root: TrieNode;

    constructor() {
        this.root = new TrieNode();
    }

    /**
     * @param {string} word
     * @return {void}
     */
    addWord(word: string): void {
        let cur = this.root;
        for (const c of word) {
            if (!cur.children.has(c)) {
                cur.children.set(c, new TrieNode());
            }
            cur = cur.children.get(c)!;
        }
        cur.isEnd = true;
    }

    /**
     * @param {string} word
     * @return {boolean}
     */
    search(word: string): boolean {
        const dfs = (node: TrieNode, i: number): boolean => {
            if (i === word.length) {
                return node.isEnd;
            }
            const c = word[i];
            if (c === '.') {
                for (const child of node.children.values()) {
                    if (dfs(child, i + 1)) {
                        return true;
                    }
                }
                return false;
            } else {
                const next = node.children.get(c);
                if (!next) {
                    return false;
                }
                return dfs(next, i + 1);
            }
        };

        return dfs(this.root, 0);
    }
}